server/coordinator/procedure/storage_impl.go

```diff
@@ -0,0 +1,118 @@
+// Copyright 2022 CeresDB Project Authors. Licensed under Apache-2.0.
+
+package procedure
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"math"
+	"path"
+
+	"github.com/CeresDB/ceresmeta/server/etcdutil"
+	"github.com/pkg/errors"
+	clientv3 "go.etcd.io/etcd/client/v3"
+)
+
+const (
+	Version              = "v1"
+	PathProcedure        = "procedure"
+	PathHistoryProcedure = "historyProcedure"
+)
+
+type EtcdStorageImpl struct {
+	client    *clientv3.Client
+	clusterID uint32
+	rootPath  string
+}
+
+func NewEtcdStorageImpl(client *clientv3.Client, clusterID uint32, rootPath string) *EtcdStorageImpl {
+	return &EtcdStorageImpl{
+		client:    client,
+		clusterID: clusterID,
+		rootPath:  rootPath,
+	}
+}
+
+// CreateOrUpdate example:
+// procedure : /{rootPath}/v1/procedure/{clusterID}/{procedureID} -> {procedureType} + {procedureState} + {data}
+func (e EtcdStorageImpl) CreateOrUpdate(ctx context.Context, meta *Meta) error {
+	str, err := encode(meta)
+	if err != nil {
+		return errors.WithMessage(err, "encode meta failed")
+	}
+	keyPath := e.generateKeyPath(meta.ID, false)
+	opPut := clientv3.OpPut(keyPath, str)
+
+	_, err = e.client.Txn(ctx).
+		Then(opPut).
+		Commit()
+
+	return err
+}
+
+// Delete example:
+// historyProcedure : /{rootPath}/v1/historyProcedure/{clusterID}/{procedureID}
+func (e EtcdStorageImpl) Delete(ctx context.Context, meta *Meta) error {
+	str, err := encode(meta)
+	if err != nil {
+		return errors.WithMessage(err, "encode meta failed")
+	}
+	keyPath := e.generateKeyPath(meta.ID, false)
+	historyKeyPath := e.generateKeyPath(meta.ID, true)
+	opDelete := clientv3.OpDelete(keyPath)
+	opPut := clientv3.OpPut(historyKeyPath, str)
+
+	_, err = e.client.Txn(ctx).Then(opDelete, opPut).Commit()
+
+	return err
+}
+
+func (e EtcdStorageImpl) Scan(ctx context.Context, batchSize int) ([]*Meta, error) {
+	metas := make([]*Meta, 0)
+	do := func(_ string, value []byte) error {
+		meta := &Meta{}
+		if err := decode(meta, string(value)); err != nil {
+			return errors.WithMessage(err, "decode meta failed")
+		}
+
+		metas = append(metas, meta)
+		return nil
+	}
+
+	startKey := e.generateKeyPath(uint64(0), false)
+	endKey := e.generateKeyPath(math.MaxUint64, false)
+
+	err := etcdutil.Scan(ctx, e.client, startKey, endKey, batchSize, do)
+	if err != nil {
+		return nil, errors.WithMessage(err, "scan procedure failed")
+	}
+	return metas, nil
+}
+
+func (e EtcdStorageImpl) generateKeyPath(procedureID uint64, isHistory bool) string {
+	var pathPrefix string
+	if isHistory {
+		pathPrefix = PathHistoryProcedure
+	} else {
+		pathPrefix = PathProcedure
+	}
+	return path.Join(e.rootPath, Version, pathPrefix, fmtID(uint64(e.clusterID)), fmtID(procedureID))
+}
+
+func fmtID(id uint64) string {
+	return fmt.Sprintf("%020d", id)
+}
+
+func encode(meta *Meta) (string, error) {
+	bytes, err := json.Marshal(meta)
```

Comment:
`Meta` contains a field whose type is ray byte, and that is to say, it may be not suitable to encode it in json format.

Reply:
`encode` function accepts a `meta` instance as a parameter. The non-utf8 data of the procedure is stored in the `rawdata` in the `meta` as `[]bytes`. Storage is only responsible for reading and writing the meta. How to convert all member variables in the `procedure` to `[]bytes` is not implemented in this PR, but is implemented by each `procedure` itself.

Reply:
This is a potential problem, we cloud resolve it by:
1. Constrain the data written in `rawdata` to be in utf8 format.
2. Replace json with proto.
I left TODO here and just resolve this conversation.

Reply:
I'm OK with replace json with proto.